[DOCS-6149] Add doc for APM instrumented against Kubernetes Service (#18768)
DataDog/documentation: content/en/tracing/guide/setting_up_apm_with_kubernetes_service.md

---
title: Setting up APM with Kubernetes Service
kind: guide
further_reading:
- link: "/containers/kubernetes/apm/"
  tag: "Documentation"
  text: "Set up trace collection"
- link: "/containers/cluster_agent/admission_controller"
  tag: "Documentation"
  text: "Admission Controller"
---

## Overview

> There are 3 main ways the Datadog Tracers send their data to the Datadog Agent for the proper handling. Those are: Host IP, Unix Domain Socket (UDS), and the Kubernetes service. These options are all setup in a way to ensure the APM data the application pod sends arrive to the Datadog Agent pod on the same node. This strategy is meant to properly balance the traffic and ensure the correct tagging of your data.

Reviewer: ```suggestion
Datadog tracers can send data to the Datadog Agent in three ways: Unix Domain Socket (UDS), host IP, or a Kubernetes service. Each option ensures that when an application pod sends APM data to the Datadog Agent pod, all of the data arrives on the same node.
This strategy is meant to properly balance traffic and ensure the correct tagging of your data.
```

Author: Pretty much added this suggestions with like 1-2 changes

> Datadog recommends to utilize the UDS strategy to send data. However, the Kubernetes service can be used as an alternative option when the `hostPath` volumes required for UDS and the `hostPort` ports required for the Host IP method are not available.

Reviewer: ```suggestion
Datadog recommends that you use UDS to send data. However, if the `hostPath` volumes required for UDS (and the `hostPort` ports required for using host IP) are not available, you can use a Kubernetes service as an alternative option.
```

## Service setup

In Kubernetes you can take advantage of [services][1] to help route traffic to downstream pods. One limitation of the default service configuration is that this routes the traffic to the downstream pods at random. In Kubernetes 1.22 the [Internal Traffic Policy feature][2] was added, giving the option to set the configuration `internalTrafficPolicy: Local` on the service. When set this directs the traffic from an application pod to the service's downstream pod *on the same node*. Maintaining that strategy of routing the APM data to the Datadog Agent on the same node.

This service is created for you automatically in our Datadog Helm Chart and Datadog Operator on Kubernetes clusters with version 1.22.0 or above. You additionally need to enable the APM port option for your Agent with the below configuration.

> 

Reviewer: I'm wondering how necessary all of this information is for the user. If you think it is very necessary to explain this background, I'd recommend putting it in a section further down on the page. For me, the only vital information here is "The service is created for you if you used the Datadog Helm chart or Operator to deploy the Agent on a Kubernetes cluster v1.22.0+. Configuration is below."

Author: I trimmed it down quite a bit to:

> In Kubernetes 1.22 the [Internal Traffic Policy feature](https://kubernetes.io/docs/concepts/services-networking/service-traffic-policy/) was added, giving the option to set the configuration internalTrafficPolicy: Local on the service. When set this directs the traffic from an application pod to the service's downstream pod on the same node.
> 
> A service for the Agent using this feature is created for you automatically in our Datadog Helm Chart and Datadog Operator on Kubernetes clusters with version 1.22.0 or above. You additionally need to enable the APM port option for your Agent with the below configuration.

> ### Agent Configuration

Reviewer: ```suggestion
### Agent configuration
```

{{< tabs >}}
{{% tab "Helm" %}}

```yaml
datadog:
  apm:
    portEnabled: true
```    

{{% /tab %}}
{{% tab "Operator" %}}

```yaml
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  #(...)
  features:
    apm:
      enabled: true
```
{{% /tab %}}
{{< /tabs >}}

## Application configuration

> To take advantage of the service the Datadog Tracer needs the `DD_AGENT_HOST` environment variable set relative to the created service. 

Reviewer: It looks like if you're using Admission Controller, you may not really interact with any kind of `DD_AGENT_HOST` env variable (seems like it's set for you) so let's not mention it here.

```suggestion
You can configure your application to use the Kubernetes service manually, or by using the [Cluster Agent Admission Controller][3].
```

### Admission controller

> The [Cluster Agent's Admission Controller][3] can simplify the configuration for APM by injecting the environment variables for connectivity. This has the configuration mode of `hostip`, `socket`, or `service` relative to the 3 options. When choosing the configuration mode of `service` the Admission Controller adds the `DD_AGENT_HOST` with the proper service name and namespace format.

Reviewer: this seems like unnecessary detail
```suggestion
```

{{< tabs >}}
{{% tab "Helm" %}}

```yaml
clusterAgent:
  admissionController:
    enabled: true
    configMode: service
```    

{{% /tab %}}
{{% tab "Operator" %}}

```yaml
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  #(...)
  features:
    apm:
      enabled: true
    admissionController:
      enabled: true
      agentCommunicationMode: service
```

{{% /tab %}}
{{< /tabs >}}

**Note:** In mixed node (Linux/Windows) environments, the Cluster Agent and its Admission Controller are relative to the Linux deployment. This may inject the wrong environment variables for the service connectivity in the Windows pods. 

### Manual configuration

> If you are setting this environment variable `DD_AGENT_HOST` manually with in your pod manifest be sure you are using the right DNS name. The general value to provide is `<SERVICE_NAME>.<SERVICE_NAMESPACE>.svc.cluster.local`. Your application pod instrumented with the Datadog Tracer should have the environment variable: 

Reviewer: ```suggestion
For manual configuration, set the environment variable `DD_AGENT_HOST` within your pod manifest, with a value of `<SERVICE_NAME>.<SERVICE_NAMESPACE>.svc.cluster.local`.
```

```yaml
    #(...)
    spec:
      containers:
      - name: "<CONTAINER_NAME>"
        image: "<CONTAINER_IMAGE>"
        env:
          - name: DD_AGENT_HOST
            value: <SERVICE_NAME>.<SERVICE_NAMESPACE>.svc.cluster.local
```

> This depends on the name of the service created for you and the namespace you have deployed the Datadog components. For example you service definition may look like:

Reviewer: ```suggestion
Replace `<SERVICE_NAME>` with the service's name, and replace `<SERVICE_NAMESPACE>` with the service's namespace.

 For example, if your service definition looks like the following:
```

```yaml
apiVersion: v1
kind: Service
metadata:
  name: datadog
  namespace: monitoring
  labels:
    #(...)
spec:
  selector:
    app: datadog
  ports:
    - protocol: UDP
      port: 8125
      targetPort: 8125
      name: dogstatsdport
    - protocol: TCP
      port: 8126
      targetPort: 8126
      name: traceport
  internalTrafficPolicy: Local
```

> In this above example the `DD_AGENT_HOST` should be set to `datadog.monitoring.svc.cluster.local`.

Reviewer: ```suggestion
Then set the value of `DD_AGENT_HOST` to `datadog.monitoring.svc.cluster.local`.

```yaml
    #(...)
    spec:
      containers:
      - name: "<CONTAINER_NAME>"
        image: "<CONTAINER_IMAGE>"
        env:
          - name: DD_AGENT_HOST
            value: datadog.monitoring.svc.cluster.local
```
```

---
title: Setting up APM with Kubernetes Service
kind: guide
further_reading:
- link: "/containers/kubernetes/apm/"
  tag: "Documentation"
  text: "Set up trace collection"
- link: "/containers/cluster_agent/admission_controller"
  tag: "Documentation"
  text: "Admission Controller"
---

## Overview

> In Kubernetes Datadog tracers can send data to the Datadog Agent in three ways: Unix Domain Socket (UDS), host IP, or a Kubernetes service. Each option ensures that when an application pod sends APM data it arrives to the Datadog Agent pod on the same node.

Reviewer: ```suggestion
In Kubernetes, Datadog tracers can send data to the Datadog Agent in three ways: Unix Domain Socket (UDS), host IP, or a Kubernetes service. Each option ensures that when an application pod sends APM data, the data arrives at a Datadog Agent pod on the same node.
```

This strategy is meant to properly balance traffic and ensure the correct tagging of your data.

Datadog recommends that you use UDS to send data. However, if the `hostPath` volumes required for UDS (and the `hostPort` ports required for using host IP) are not available, you can use a Kubernetes service as an alternative option.

## Service setup

In Kubernetes 1.22 the [Internal Traffic Policy feature][1] was added, giving the option to set the configuration `internalTrafficPolicy: Local` on the service. When set this directs the traffic from an application pod to the service's downstream pod *on the same node*.

> A service for the Agent using this feature is created for you automatically in our Datadog Helm Chart and Datadog Operator on Kubernetes clusters with version 1.22.0 or above. You additionally need to enable the APM port option for your Agent with the below configuration.

Reviewer: ```suggestion
A service for the Agent using this feature is created for you automatically by the Datadog Helm chart and Datadog Operator on clusters with Kubernetes v1.22.0+. You additionally need to enable the APM port option for your Agent with the below configuration.
```
